Sprint_14

Wilwous opened the pull request:
Добавлена функциональность создания и отмены трекера.
Реализована смена состояния кнопки при готовности создания трекера.
Добавлены ассеты цветов, логотипа и лаунчскрина для улучшения внешнего вида приложения.
Создан TabBarViewController с вкладками "Трекеры" и "Статистика" для более удобной навигации.
Реализован TrackerViewController для отображения трекеров с функциями поиска и выбора даты.
Добавлен StatisticsViewController для удобного просмотра статистики.
Введены структуры Tracker, TrackerCategory и TrackerRecord для хранения данных о трекерах и записях о выполнении.
Создан экран создания трекеров и TimetableCreation для настройки расписания.
Введен алерт для информирования пользователя об ошибке при отсутствии названия трекера.

Commit:
Добавил ассеты цветов и лого.
Сверстал лаунчскрин.

Commit:
feat: Добавлен TabBarViewController

- Добавлены вкладки "Трекеры" и "Статистика" в TabBarViewController
- Добавлена разделяющая линия

feat: Добавлен TrackerViewController

- Реализован поиск в TrackerViewController
- Добавлена возможность выбора даты
- Добавлена кнопка для добавления трекеров
- Добавлены лейблы для различных элементов в TrackerViewController

feat: Добавлен StatisticsViewController

chore: Добавлена иконка приложения и прочие ассеты.

Commit:
fix: Исправление и доработка форматирования даты

- Добавлены констрейны ширины для корректного отображения форматирования даты в цифрах
- Удалена дублирующая строка кода

Commit:
feat: Добавлены структуры Tracker, TrackerCategory и TrackerRecord

- Создана сущность Tracker для хранения информации о трекере с уникальным идентификатором, названием, цветом, эмоджи и расписанием.
- Введена TrackerCategory, представляющая категорию трекеров с заголовком и массивом трекеров внутри.
- Добавлена TrackerRecord для хранения записей о выполнении трекера на определенную дату с id трекера и датой выполнения.

feat: Обновлены свойства и методы в TrackersViewController

- Добавлены свойства categories и completedTrackers для хранения списка категорий и выполненных трекеров соответственно.
- Реализован механизм добавления новых трекеров в категории и отметок о выполнении в completedTrackers.

Commit:
fix

Commit:
feat: Добавлена функциональность кнопки добавления трекера

- Добавлена приватная функция addTrackerButtonTapped для обработки нажатия на кнопку добавления трекера.
- Реализован механизм перехода к экрану создания трекера при нажатии на кнопку.

Добавлен экран для создания трекеров

- Создан экран TrackerCreator для создания новых трекеров.
- Добавлены элементы интерфейса: метка "Создание трекера", кнопки "Привычка" и "Нерегулярное событие".
- Реализованы констрейнты для размещения элементов интерфейса на экране.

chore: Переименована переменная для лучшей читаемости

Commit:
UPD:
feat: Добавлена функциональность кнопки добавления трекера
- Добавлена приватная функция addTrackerButtonTapped для обработки нажатия на кнопку добавления трекера.
- Реализован механизм перехода к экрану создания трекера при нажатии на кнопку.

Добавлен экран для создания трекеров

- Создан экран TrackerCreator для создания новых трекеров.
- Добавлены элементы интерфейса: метка "Создание трекера", кнопки "Привычка" и "Нерегулярное событие".
- Реализованы констрейнты для размещения элементов интерфейса на экране.

chore: Переименована переменная для лучшей читаемости

Commit:
feat: Добавлен экран создания привычек

- Создан экран HabitCreation для создания новых привычек.
- Добавлены элементы интерфейса: метка "Новая привычка", текстовое поле для ввода названия привычки, сообщение об ограничении символов, таблица для выбора категории и расписания, кнопки "Отменить" и "Создать".
- Реализованы констрейнты и настройка интерфейса для корректного отображения элементов на экране.
- Реализована таблица с ячейками для выбора категории и расписания привычки.

chore: Переорганизация структуры проекта

Commit:
feat: Добавлен экран создания расписания

- Создан класс TimetableCreation для отображения экрана создания расписания.
- Реализован выбор дней недели с помощью UITableView.
- Добавлена кнопка "Готово" для завершения выбора дней недели.
- Добавлен протокол TimetableCreationDelegate для обратной связи с родительским контроллером.
- Создан класс TimetableTableView для ячейки таблицы с днями недели.
- Добавлена возможность выбора дней недели с помощью UISwitch в ячейке таблицы.

Commit:
feat: Добавлена коллекция трекера.
- Реализовано отображение заглушки.
- Добавлена возможность отметки трекера для пометки выполненных задач.
- Реализовано всплывающее сообщение о ограничение на количество символов.
- Добавлена функциональность фильтрации по дате и поиску для удобства работы с данными.

refactor:
- Переименованы переменные для более ясного понимания их назначения.
- Добавлены комментарии Mark для улучшения навигации и структурирования кода.
- Организован код и файлы для улучшения читаемости и поддерживаемости проекта.
- Пересмотрены и скорректированы констрейнты для улучшения внешнего вида интерфейса.

Commit:
feat: Добавлена функциональность создания и отмены трекера.
Введен алерт для информирования пользователя об ошибке при отсутствии названия трекера.
Реализована смена состояния кнопки при готовности создания трекера.

Commit:
fix: Поправил смену состояние кнопки создать.

So1ovyov commented on Tracker/Helpers/DataSource.swift:
```diff
@@ -0,0 +1,35 @@
+
+import UIKit
+
+final class DataSource {
```
Круто, что моковые данные вынес в отдельную сущность, но кажется можно было сделать проще и с меньшим количеством кода:
```
enum MockData {
    
    static let trackerCategories: [TrackerCategory] = [
        TrackerCategory(
            headline: "Домашний уют",
            trackers: [
                Tracker(id: UUID(), name: "Поливать цветы", color: "colorSelection16", emoji: "🪴", timetable: [.monday,.wednesday,.friday], completedDays: []),
                Tracker(id: UUID(), name: "Налить стакан пива", color: "colorSelection11", emoji: "🍺", timetable: [.friday,.saturday,.sunday], completedDays: [])
            ]
        ),
        TrackerCategory(
            headline: "Уборка",
            trackers: [
                Tracker(id: UUID(), name: "Помыть посуду", color: "colorSelection12", emoji: "🧽", timetable: [.saturday], completedDays: []),
                Tracker(id: UUID(), name: "Помыть полы", color: "colorSelection14", emoji: "🪣", timetable: [.tuesday, .sunday], completedDays: [])
            ]
        )
    ]
}

```

So1ovyov replied:
И тогда не нужен класс, синглтон и метод

So1ovyov commented on Tracker/Helpers/UIColor + Extension.swift:
```diff
@@ -0,0 +1,39 @@
+import UIKit
+
+extension UIColor {
+    static let colorMap: [String: UIColor] = [
```
Рекомендую всегда использовать библиотеки генерации ресурсов типа [SwiftGen](https://github.com/SwiftGen/SwiftGen).
Избавит от многих проблем при работе с ресурсами(цветами, строками, картинками и прочим).

So1ovyov commented on Tracker/Helpers/WeekDay.swift:
```diff
@@ -0,0 +1,58 @@
+
+import Foundation
+
+enum WeekDay: String, CaseIterable {
```
Отличное решение

So1ovyov commented on Tracker/TabBarManager/StatisticsManager/StatisticsViewController.swift:
```diff
@@ -0,0 +1,73 @@
+    
+    // MARK: - Setup Methods
+    private func addElemens() {
+        view.addSubview(emptyStateImageView)
```
Обрати внимание на инструмент `forEach` при верстке и настройке элементов, он поможет тебе сократить повторяющийся код
Вот пример:
```
        [view1, 
         view2,
         view3].forEach {
            $0.translatesAutoresizingMaskIntoConstraints = false
             view.addSubview($0)
        }
```
Таким образом все строки `translatesAutoresizingMaskIntoConstraints` и `addSubview` выше можно удалить и очень сильно сократить код

So1ovyov commented on Tracker/TrackerManager/Timetable/TimetableCreation.swift:
```diff
@@ -0,0 +1,149 @@
+        tableView.backgroundColor = .ypBackgroundDay
+        tableView.layer.cornerRadius = 16
+        tableView.translatesAutoresizingMaskIntoConstraints = false
+        tableView.register(TimetableTableView.self,
```
А может так лучше переносить ?
Меньше места в ширину )
```
        tableView.register(
            TimetableTableView.self,
            forCellReuseIdentifier: TimetableTableView.cellID
        )
```

So1ovyov commented on Tracker/TrackerManager/View/TrackerCollection.swift:
```diff
@@ -0,0 +1,235 @@
+    
+    // MARK: - Setup View
+    private func addElements() {
+        contentView.addSubview(topBackgroundView)
```
В такие кейсы `forEach` так и просится, попробуй использовать.

So1ovyov commented on Tracker/TrackerManager/View/TrackerViewController.swift:
```diff
@@ -0,0 +1,378 @@
+    }
+    
+    private func updateCategoriesView() {
+        if visibleCategories.isEmpty {
```
Давай сократим метод с помощью тернарного оператора:
```
    private func updateCategoriesView() {
        trackersViewStubs.isHidden = !visibleCategories.isEmpty
        trackerCollectionView.backgroundView = visibleCategories.isEmpty ? trackersViewStubs : nil
    }
```

So1ovyov commented on Tracker/TrackerManager/View/TrackersViewStubs.swift:
```diff
@@ -0,0 +1,59 @@
+    
+    // MARK: - Setup View
+    private func addElements() {
+        self.addSubview(willTrackLabel)
```
А почему через `self` ? Можно же и без него, проверь проект на наличие такого кейса и исправь по возможности
```
    private func addElements() {
        addSubview(willTrackLabel)
        addSubview(imagesViewStub)
    }
```

Commit:
fix: - Устранена проблема с нажатием по таблице в выборе расписания трекера
- Исправлена проблема с ассетами: теперь появляется галочка завершения трекера и заглушка при отсутствии трекера
- Исправлена проблема, при которой страница создания трекера открывалась сразу
- Улучшена фильтрация поиска по выбранному дню
- Исправлено расположение кнопок создания и отмены трекера
- Улучшен функционал создания нового трекера

refactor: - Заменен поиск трекера с использованием UISearchController на UISearchBar
- Проведена реструктуризация кода для сокращения его объема

Commit:
fix: - Поправил ассеты
- Поправил нажатие на таблицу

Commit:
refactor: Убрал лишний код

Wilwous commented:
-Устранена проблема с нажатием по таблице в выборе расписания трекера.
- Исправлена проблема с ассетами: теперь появляется галочка завершения трекера и заглушка при отсутствии трекера.
- Исправлена проблема, при которой страница создания трекера открывалась сразу.
- Улучшена фильтрация поиска по выбранному дню.
- Исправлено расположение кнопок создания и отмены трекера.
- Улучшен функционал создания нового трекера.
- Переписаны участки кода с учетом рекомендаций по стилю и эффективности.

So1ovyov commented:
@Wilwous проделана большая работа, проект получился качественным, код аккуратный и понятный. Молодец!
